clippy_lints/src/mut_key.rs

```diff
@@ -19,10 +19,14 @@ declare_clippy_lint! {
     /// so having types with interior mutability is a bad idea.
     ///
     /// ### Known problems
-    /// It's correct to use a struct, that contains interior mutability
-    /// as a key, when its `Hash` implementation doesn't access any of the interior mutable types.
-    /// However, this lint is unable to recognize this, so it causes a false positive in theses cases.
-    /// The `bytes` crate is a great example of this.
+    /// It's correct to use a struct that contains interior mutability as a key, when its
+    /// implementation of `Hash` or `Ord` doesn't access any of the interior mutable types.
+    /// However, this lint is unable to recognize this, so it will often cause false positives in
+    /// theses cases.  The `bytes` crate is a great example of this.
+    ///
+    /// Additionally, if a struct/enum contains interior mutability behind indirection then this
+    /// lint is unlikely to catch it.  For example, `struct BadKey<'a>(&'a Cell<usize>)` will not
+    /// be detected by this lint.
```

Comment:
```suggestion
    /// Additionally, if a struct/enum contains a field with interior mutability, then this
    /// lint will not catch it.  For example, `struct BadKey<'a>(&'a Cell<usize>)` will not
    /// be detected by this lint.
```

Reply:
> Additionally, if a struct/enum contains _a field with interior mutability_, then this lint will not catch it

This isn't strictly true: what matters is the indirection because that's what impacts the `Freeze` trait, which is what the `Adt` case checks.  For example:

This will lint, because the bytes of `BadKey` can change without a mutable reference (i.e. `BadKey` is not `Freeze`):
```rust
struct BadKey {
    uh_oh: Cell<usize>,
}
```
but this won't (despite being just as bad) because the mutability happens to memory _owned_ by `SilentlyBadKey`, not `SilentlyBadKey` itself:
```rust
struct SilentlyBadKey {
    uh_oh: Box<Cell<usize>>,
}
```

The other change is correct, though - I'll fix that now.

Reply:
Good point. How about "a field with interior mutability behind a pointer like `Box<Cell<usize>>`"?